File: authentication/auth_example.js
var crypto = require('crypto')
var request = require('request') // request package installed from npm
var uuidv4 = require('uuid/v4') // uuid package installed from npm

// Authentication details
API_KEY = 'YOUR_API_KEY' // From the developer portal
API_SECRET = 'YOUR_API_SECRET' // From the developer portal

// The request method and endpoint
API_URL = 'https://api-sandbox.bitpesa.co/v1/senders'
API_METHOD = 'POST'

// Request-specific data
nonce = uuidv4() // Must be unique per request
body = JSON.stringify({
  sender: {
    country: 'UG',
    phone_country: 'UG',
    phone_number: '752403639',
    email: 'email@domain.com',
    first_name: 'Example',
    last_name: 'User',
    city: 'Kampala',
    street: 'Somewhere 17-3',
    postal_code: '798983',
    birth_date: '1970-01-01',
    documents: [
      {
        upload:
          'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAIAAACQd1PeAAAACXBIWXMAAAsT\nAAALEwEAmpwYAAAAB3RJTUUH4gEeCTEzbKJEHgAAAB1pVFh0Q29tbWVudAAA\nAAAAQ3JlYXRlZCB3aXRoIEdJTVBkLmUHAAAADElEQVQI12P4z8AAAAMBAQAY\n3Y2wAAAAAElFTkSuQmCC',
        upload_file_name: 'passport.png',
        metadata: { meta: 'data' }
      }
    ],
    ip: '127.0.0.1',
    metadata: { meta: 'data' }
  }
})

// Create hash of body data
var hash = crypto.createHash('sha512')
hash.update(body)

// Generate the signature
toSign = [
  nonce,
  API_METHOD,
  API_URL,
  hash.digest('hex')
].join('&')

var hmac = crypto.createHmac('sha512', API_SECRET)
hmac.update(toSign)
auth_signature = hmac.digest('hex')

// Build the request headers
headers = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
  'Authorization-Key': API_KEY,
  'Authorization-Nonce': nonce,
  'Authorization-Signature': auth_signature
}

// Send the request
var options = {
  url: API_URL,
  method: API_METHOD,
  headers: headers,
  body: body
}

request(options, function(error, res, body) {
  if (error) {
    // handle error
    console.log(error)
  } else {
    // response received
    console.log(body)
  }
})

// Expected response body
{"object":{"id":"dd9e77f5-5727-42a9-b91d-fdded7edbeac","type":"person","state":"initial","country":"UG","street":"Unknown 17-3","postal_code":"798983","city":"Kampala","phone_country":"UG","phone_number":"752403639","email":"sender@bitpesa.demo","ip":"127.0.0.1","address_description":null,"first_name":"Johnny","last_name":"English","birth_date":"1970-01-01","documents":[{"id":"7ae470e5-743f-4ef7-8af5-2ab00fcdb966","upload":"https://bitpesa-development-documents.s3-eu-west-1.amazonaws.com/documents/uploads/7ae470e5-743f-4ef7-8af5-2ab00fcdb966/thumbnail/passport.png?X-Amz-Algorithm=AWS4-HMAC-SHA256\\u0026X-Amz-Credential=AKIAI3CWSFDBPPJHAYYQ%2F20180130%2Feu-west-1%2Fs3%2Faws4_request\\u0026X-Amz-Date=20180130T100152Z\\u0026X-Amz-Expires=3600\\u0026X-Amz-SignedHeaders=host\\u0026X-Amz-Signature=9e3a4630e4622ff8667a82aaf501fcea5e3b1300606424453cf91ca212af5ff8","metadata":{"meta":"data"},"upload_file_name":"passport.png","upload_content_type":"image/png","upload_file_size":152}],"metadata":{"my":"data"},"providers":{"NGN::Bank":"paga"}}}

File: authentication/auth_example.cs
using System;
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.IO;
using System.Collections.Specialized;

namespace transaction
{
	class MainClass
	{
		public static void Main(string[] args)
		{
			// Authentication details
			var apiKey = "YOUR_API_KEY";
			var apiSecret = "YOUR_API_SECRET";
			
			// The request method and endpoint
			var url = "https://api-sandbox.bitpesa.co/v1/senders";
			var requestMethod = "POST";
			
			// Request specific data
			var nonce = System.Guid.NewGuid().ToString(); // Must be unique per request
			var body = "{\"sender\":{\"country\":\"UG\",\"phone_country\":\"UG\",\"phone_number\":\"752403639\",\"email\":\"example@home.org\",\"first_name\":\"Johnny\",\"last_name\":\"English\",\"city\":\"Kampala\",\"street\":\"Unknown 17-3\",\"postal_code\":\"798983\",\"birth_date\":\"1900-12-31\",\"ip\":\"127.0.0.1\",\"metadata\":{\"my\":\"data\"}}}";

			// Create hash of body data
			string md5HashOfBody = GetSHA512Hash(body);
			
			// Generate the signature
			var toSign = string.Format("{0}&{1}&{2}&{3}", nonce, requestMethod, url, md5HashOfBody);
			var authSignature = GetHMACHash(apiSecret, toSign);

			// Build the request headers
			HttpWebRequest req = (HttpWebRequest)WebRequest.Create(url);
			req.Accept = "application/json";
			req.ContentType = "application/json";
			req.Headers.Add("Authorization-Key", apiKey);
			req.Headers.Add("Authorization-Nonce", nonce);
			req.Method = "POST";
			req.Headers.Add("Authorization-Signature", authSignature);
			byte[] bytes = Encoding.UTF8.GetBytes(body);
			req.ContentLength = bytes.Length;

			// Send the request
			Stream requestStream = req.GetRequestStream();
			requestStream.Write(bytes, 0, bytes.Length);

			WebResponse response = req.GetResponse();
			Stream stream = response.GetResponseStream();
			StreamReader reader = new StreamReader(stream);

			var result = reader.ReadToEnd();
			stream.Dispose();
			reader.Dispose();

			Console.WriteLine(result);
		}

		static string GetSHA512Hash(string input)
		{
			SHA512 digest = SHA512.Create();
			// Convert the input string to a byte array and compute the hash.
			byte[] data = digest.ComputeHash(Encoding.UTF8.GetBytes(input));

			// Create a new Stringbuilder to collect the bytes
			// and create a string.
			StringBuilder sBuilder = new StringBuilder();

			// Loop through each byte of the hashed data 
			// and format each one as a hexadecimal string.
			for (int i = 0; i < data.Length; i++)
			{
				sBuilder.Append(data[i].ToString("x2"));
			}

			// Return the hexadecimal string.
			return sBuilder.ToString();
		}

		static string GetHMACHash(string api_secret, string input)
		{
			HMACSHA512 hmacDigest = new HMACSHA512(Encoding.UTF8.GetBytes(api_secret));
			// Convert the input string to a byte array and compute the hash.
			byte[] data = hmacDigest.ComputeHash(Encoding.UTF8.GetBytes(input));

			// Create a new Stringbuilder to collect the bytes
			// and create a string.
			StringBuilder sBuilder = new StringBuilder();

			// Loop through each byte of the hashed data 
			// and format each one as a hexadecimal string.
			for (int i = 0; i < data.Length; i++)
			{
				sBuilder.Append(data[i].ToString("x2"));
			}

			// Return the hexadecimal string.
			return sBuilder.ToString();
		}
	}
}
